apps/sms/views/shared/test/unit/system_message_handler_test.js

```diff
@@ -0,0 +1,778 @@
+
+      // When app is run as activity we should not listen for 'sms-received' and
+      // 'notification' system messages.
+      sinon.assert.notCalled(navigator.mozSetMessageHandler);
+    });
+
+    test('if app is run in non-activity mode', function() {
+      ActivityShim.hasPendingRequest.returns(false);
+
+      SystemMessageHandler.init();
+
+      sinon.assert.calledTwice(window.navigator.mozSetMessageHandler);
+      sinon.assert.calledWith(
+        window.navigator.mozSetMessageHandler, 'sms-received'
+      );
+      sinon.assert.calledWith(
+        window.navigator.mozSetMessageHandler, 'notification'
+      );
+    });
+  });
+
+  suite('sms-received system message', function() {
+    var appStub, wakeLockStub, notificationStub;
+
+    function assertWakeLockIsRequested() {
+      sinon.assert.calledOnce(navigator.requestWakeLock);
+      sinon.assert.calledWith(navigator.requestWakeLock, 'cpu');
+
+      // We shouldn't release wake lock until we completely handled new message.
+      sinon.assert.notCalled(wakeLockStub.unlock);
+    }
+
+    function assertWakeLockIsReleased() {
+      sinon.assert.calledOnce(wakeLockStub.unlock);
+    }
+
+    function assertUserIsNotified() {
+      sinon.assert.calledOnce(Notify.ringtone);
+      sinon.assert.calledOnce(Notify.vibrate);
+    }
+
+    function assertUserIsNotNotified() {
+      sinon.assert.notCalled(Notify.ringtone);
+      sinon.assert.notCalled(Notify.vibrate);
+    }
+
+    setup(function() {
+      appStub = sinon.stub({ launch() {} });
+
+      wakeLockStub = sinon.stub({
+        unlock: () => {}
+      });
+
+      notificationStub = sinon.stub({
+        addEventListener() {},
+        close() {}
+      });
+
+      this.sinon.stub(navigator.mozApps, 'getSelf').returns(
+        Promise.resolve(appStub)
+      );
+      this.sinon.stub(navigator, 'requestWakeLock').returns(wakeLockStub);
+      this.sinon.spy(Notify, 'ringtone');
+      this.sinon.spy(Notify, 'vibrate');
+      this.sinon.stub(SilentSms, 'checkSilentModeFor').returns(
+        Promise.resolve(false)
+      );
+      this.sinon.stub(Navigation, 'isCurrentPanel').returns(false);
+      this.sinon.stub(Navigation, 'toPanel').returns(Promise.resolve());
+      this.sinon.stub(NotificationHelper, 'send').returns(notificationStub);
+
+      this.sinon.stub(NotificationHelper, 'getIconURI').returns(null);
+      NotificationHelper.getIconURI.withArgs(appStub).returns('iconURI');
+
+      this.sinon.stub(Contacts, 'findByAddress').returns(
+        Promise.resolve([])
+      );
+    });
+
+    suite('received message from silenced number', function() {
+      setup(function() {
+        SilentSms.checkSilentModeFor.returns(Promise.resolve(true));
+      });
+
+      test('does not notify user in any way', function(done) {
+        var handlingPromise = SystemMessageHandler.onSmsReceivedSystemMessage(
+          MockMessages.sms()
+        );
+
+        assertWakeLockIsRequested();
```

Comment:
I test for the correct wake lock request/release only in a subset of tests where it makes sense and tried to keep other tests simpler without this check :)